### DISCUSSION TITLE
simple trick to immediately get better at `nu` in one-minute

### DISCUSSION THREAD

**Discussion by goyalyashpal** [nushell/nushell#12318] [Show and tell]
# Simple trick to get better at `nu` in one minute
2024-03-28 Guruvaar

Sorry for the catchy title, I just had to write it lol 😄. However, to
stay true to it jump to [Experience to the rescue] as TL;DR.

[Experience to the rescue]: #yq-experience-and-nu-documentation-to-the-rescue

## Past-time with nu

The last time I was using [`nu`], on seeing the manipulations beautifully
transforming the tabular data, the deeply nested hierarchies of `help` etc;
I remember thinking that it would be nice to be able to see what the
input should be to get that table directly.

Also for the reason that often times, different types of data ended up
getting rendered the same.

But I couldn't figure out how to do it back then, so I passed that thought.
Eventually I took hiatus from using `nu`.

## Using file processors to make sense

Recently, I installed a whole slew of new tools, thanks to the awesome
[`scoop`] community powered, UNLICENSE-licensed CLI package manager.
One of them is the awesome [`yq`], which allows to convert different
types of input data to `yaml` or `json`, among others.

And it turns out that JSON is surprisingly unambiguous and easy to
grasp format. I came to appreciate this beauty of JSON only after
I started using other data formats on daily basis.

Converting a `pyproject.toml` to json using
`yq --input-format toml --output-format json` (or `-p toml -o json` for
short) made me immediately understand the `toml` format too, whereas I
always used to forget its specifics before.

This brought me into habbit of piping any compatible data into `yq -p <type>`
and having it converted into format of my choice: `yaml` for brevity, or
`json` for exploring untouched waters, or piping it to query with [`jq`].

[`nu`]: https://github.com/nushell/nushell "A modern [Rust-based] shell where all data is structured!"
[`scoop`]: https://github.com/ScoopInstaller/Scoop "A command-line [package manager] for Windows"
[`yq`]: https://mikefarah.gitbook.io/yq "Portable command-line [data file] processor"
[`jq`]: https://jqlang.github.io/jq "Lightweight and flexible command-line JSON processor"
[`header-on-separator`]: https://www.nushell.sh/blog/2023-08-22-nushell_0_84_0.html#new-options-for-more-compact-table-display-zhiburt-9796 "Compact table display"

## Flashback to nu

After `yq` and `jq`, I got back to using `nu`, thanks again to ease of
updating using [`scoop`]. Of course, I liked the premise of `nu` and
loved the mesmerizing rounded tables with [`header-on-separator`] (did
I already say that 😂).

I was trying to replace one output column of a command with some other
data. Specifically, "type" from nu's `ls` (hereafter: old-data) with
"type" from astron/darwinsys' `file` as the latter detects some
filetypes better. I was able to get a `name: type` form of yaml-like
output from `file` (hereafter: new-data).

```console
> ls -m *.scf | reject modified
╭─#─┬─────name──────┬──type───┬───size───╮
│ 0 │ shortcuts.scf │ unknown │ 28.9 KiB │
╰───┴───────────────┴─────────┴──────────╯

> file *.scf
shortcuts.scf: XML 1.0 document, ASCII text, with CRLF line terminators

> env ls -A1 | file --mime-type -f -
pixmaps.txt:      text/plain
session.ini:      application/x-wine-extension-ini
shortcuts.scf:    text/xml
```

... but couldn't figure out how to convert it to `nu`'s table required
for `merge`ing into nu's `ls` output. So, naively I tried to add a
header row to the new-data in raw form, and fiddling with things like
`wrap`. But no go hereafter.

```console
> # ... | from yaml
> ... | lines | prepend "name: type" | str join (char newline) | from yaml
╭──────────────────┬──────────────────────────────────╮
│ name             │ type                             │
│ pixmaps.txt      │ text/plain                       │
│ session.ini      │ application/x-wine-extension-ini │
│ shortcuts.scf    │ text/xml                         │
╰──────────────────┴──────────────────────────────────╯

```

## yq Experience and nu documentation to the rescue

On one hand, I was thinking about how to represent this issue in 
[MWE]. On the other, I hesitantly looked around searching the [web][show custom column in nu ls] 
and stumbled upon this article in nu's documentation: [cheat_sheat] 
going through various data types, and `to yaml` command etc.

Due to both these directions, it incurred to me to examine their
underlying representation in a more explicit and common-baseline format.

[MWE]: . "Minimum Working Example"
[show custom column in nu ls]: https://duckduckgo.com/?q=show+custom+column+in+nu+ls&t=vivaldi&ia=web "ddg web search"
[working_with_tables]: https://www.nushell.sh/book/working_with_tables.html "Working with tables"
[vars_and_exprs]: https://www.nushell.sh/book/variables_and_subexpressions.html "Variables and Subexpressions"
[cheat_sheat]: https://www.nushell.sh/book/cheat_sheet.html "Nushell cheat sheet"

```console
> # ... | from yaml | to nuon --indent 2
> ... | from yaml | to json
{
  "pixmaps.txt": "text/plain",
  "session.ini": "application/x-wine-extension-ini",
  "shortcuts.scf": "text/xml"
}

> ls | reject modified | to json
[
  {
    "name": "pixmaps.txt",
    "type": "file",
    "size": 582
  },
  ...
]
```

Ohw snapp! The "new-data" is essentially in pivot-table's form. I've to
just `transpose` each json object. Yep, the following output matches with 
underlying format of "old-data". And voilla, the rendering confirms too 😇.

```console
> ... | from yaml | transpose name type | to json
[
  {
    "name": "pixmaps.txt",
    "type": "text/plain"
  },
  ...
]

> ... | from yaml | transpose name type
╭─#─┬─────name──────┬───────────────type───────────────╮
│ 0 │ pixmaps.txt   │ text/plain                       │
 ...
```

Now, it is fairly standar to merge these two.
* With `join`, you may want to additionaly `reject <old-column>` before
joining.
* Unlike `merge`, `join` allows merging in either direction (good).
* With `merge`, you've to ensure that order of the "old-data" and
"new-data" is same, as overwrites index-wise, not per the matching
key-value (inferior for flexibility).

```console
> # ls -am | join --inner (env ls -A1 | file --mime-type -f - | from yaml | transpose name type) name
> env ls -A1 | file --mime-type -f - | from yaml | transpose name type | join --inner (ls -am) name
╭─#─┬─────name──────┬───────────────type───────────────┬───type_────┬───size───╮
│ 0 │ pixmaps.txt   │ text/plain                       │ text/plain │    582 B │
│ 1 │ session.ini   │ application/x-wine-extension-ini │ text/plain │  2.1 KiB │
│ 2 │ shortcuts.scf │ text/xml                         │ unknown    │ 28.9 KiB │
╰───┴───────────────┴──────────────────────────────────┴────────────┴──────────╯

> ls -am | merge (env ls -A1 | file --mime-type -f - | from yaml | transpose name type)
╭─#─┬─────name──────┬───────────────type───────────────┬───size───╮
│ 0 │ pixmaps.txt   │ text/plain                       │    582 B │
│ 1 │ session.ini   │ application/x-wine-extension-ini │  2.1 KiB │
│ 2 │ shortcuts.scf │ text/xml                         │ 28.9 KiB │
╰───┴───────────────┴──────────────────────────────────┴──────────╯

```

### Conclusion

* If on windows, use [`scoop`] package manager to manage app installations.
* Get familiar with community based tools for structured data
processing, conversion & querying like [`yq`], [`jq`], [`nu`].
* Experience with `sql` will help in all these.
* To get better grasp of data, convert it to `json` for understanding,
`yaml` for brevity
* Thinking about MWEss often help too

<details>
<summary>
<h3>Versions at the time of writing</h3>
</summary>

```console
> scoop --version | head -n2
Current Scoop version:
v0.3.1 - Released at 2022-11-15

> nu --version
0.91.0

> yq --version
yq (https://github.com/mikefarah/yq/) version v4.42.1

> jq --version
jq-1.7.1

> sqlite3 --version
3.45.2 2024-03-12 11:06:23

> systeminfo | grep -iA1 "os name" | from yaml
╭────────────┬───────────────────────────────────────────╮
│ OS Name    │ Microsoft Windows 10 Home Single Language │
│ OS Version │ 10.0.19045 N/A Build 19045                │
╰────────────┴───────────────────────────────────────────╯
```

</details>

**Comment by Adeor** [comment]
To add to the discussion, [jc](https://github.com/kellyjonbrazil/jc) (also linked in [awesome-nu](https://github.com/nushell/awesome-nu?tab=readme-ov-file#integrations)) is also really nice to convert data into nu data structures via json as IR.

It can even be used to semi-automatically shadow supported built-ins. I current do so like this: 

<details>
  <summary>scripts/bin/gen-aliases.nu</summary>

``` nushell
#!/usr/bin/env nu

use std/log

# Has to be done via a generator rn
# See https://github.com/nushell/nushell/issues/5068 for details


# Generates a function (without testing, if it works)
def gen-function [name :string, content: string, file :path] {
  $"def '($name)' [] { ($content) }\n" | save -a $file
  return $name
}

# Generate an alias (without testing, if it works)
def gen-alias [alias :string, target: string, file :path] {
  $"alias '($alias)' = ($target)\n" | save -a $file
}

# Generates a function, if it can find the exec
def maybe-gen-function [exec :string, args: string, name: string, file :path] {
  let $content = $"^($exec) ($args)"

  if (which $exec | is-empty) {
    log warning $"($exec) isn't in path. The function ($name) won't be generated."
  } else {
    gen-function $name $"^($exec) ($args)" $file
    gen-alias $"($name) --help" $"^($exec) --help"  $file
    gen-alias $"($name) -h" $"^($exec) -h" $file
  }
}

# So it's easier to call from a record
def maybe-gen-function-jc [exec :string, file: path, eargs :string, jcargs :string, fname :string] {
  maybe-gen-function $exec $"($eargs) | jc ($jcargs) | from json" $fname $file
}

# Shadow the built-ins for the functions I use
let file = $"($nu.default-config-dir)/aliases.nu" 
if ($file | path exists) { rm $file }

[
  #{ exec: "acpi" }
  #{ exec: "airport" }
  #{ exec: "airport-s" }
  #{ exec: "amixer" }
  #{ exec: "apt-cache-show" }
  #{ exec: "apt-get-sqq" }
  #{ exec: "arp" }
  #{ exec: "asciitable" }
  #{ exec: "asciitable-m" }
  #{ exec: "blkid" }
  #{ exec: "bluetoothctl" }
  #{ exec: "cbt" }
  #{ exec: "cef" }
  #{ exec: "cef-s" }
  #{ exec: "certbot" }
  #{ exec: "chage" }
  #{ exec: "cksum" }
  #{ exec: "clf" }
  #{ exec: "clf-s" }
  { exec: "crontab" }
  { exec: "crontab-u" }
  #{ exec: "csv" }
  #{ exec: "csv-s" }
  { exec: "curl", eargs: "--head", jcargs: "curl-head", fname: "curl --head" } # nu has `http head`, but for some advanced use-cases it's nice
  #{ exec: "date" } # already in nushell
  #{ exec: "datetime-iso" }
  #{ exec: "debconf-show" }
  #{ exec: "df" } # already in nushell
  #{ exec: "dig" }
  #{ exec: "dir" }
  #{ exec: "dmidecode" }
  #{ exec: "dpkg-l" }
  #{ exec: "du" } # already in nu
  #{ exec: "efibootmgr" }
  #{ exec: "email-address" }
  #{ exec: "env" }  # already in nu (via $env)
  { exec: "ethtool" }
  { exec: "file" }
  #{ exec: "find" } # I use fd
  #{ exec: "findmnt" }
  #{ exec: "finger" }
  { exec: "free" }
  #{ exec: "fstab" }
  #{ exec: "git-log" } # hmm
  #{ exec: "git-log-s" } # hmm
  #{ exec: "git-ls-remote" } # hmm
  { exec: "gpg" }
  #{ exec: "group" }
  #{ exec: "gshadow" }
  #{ exec: "hash" } # already in nu
  #{ exec: "hashsum" }
  #{ exec: "hciconfig" }
  #{ exec: "history" } # already in nu
  #{ exec: "host" } # already in nu as `sys host`
  #{ exec: "hosts" }
  #{ exec: "http-headers" }
  { exec: "id" }
  #{ exec: "ifconfig" }
  #{ exec: "ini" }
  #{ exec: "ini-dup" }
  { exec: "iostat" }
  { exec: "iostat-s" }
  #{ exec: "ip-address" }
  #{ exec: "ipconfig" }
  #{ exec: "iptables" }
  #{ exec: "ip" eargs: "route", jcargs: "ip-route", aname: "ip route"}
  #{ exec: "iw-scan" }
  #{ exec: "iwconfig" }
  #{ exec: "jar-manifest" }
  #{ exec: "jobs" } # already in nu as `job list`
  #{ exec: "jwt" }
  #{ exec: "kv" }
  #{ exec: "kv-dup" }
  #{ exec: "last" }
  #{ exec: "ls" } # already in nu
  #{ exec: "ls-s" } # already in nu
  { exec: "lsattr" }
  #{ exec: "lsb-release" }
  { exec: "lsblk" }
  { exec: "lsmod" }
  { exec: "lsof" }
  { exec: "lspci" }
  { exec: "lsusb" }
  #{ exec: "m3u" }
  #{ exec: "mdadm" }
  #{ exec: "mount" }
  #{ exec: "mpstat" }
  #{ exec: "mpstat-s" }
  #{ exec: "needrestart" }
  #{ exec: "netstat" }
  #{ exec: "nmcli" }
  #{ exec: "nsd-control" }
  #{ exec: "ntpq" }
  #{ exec: "openvpn" }
  #{ exec: "os-prober" }
  #{ exec: "os-release" }
  { exec: "pacman" }
  { exec: "passwd" }
  #{ exec: "path" }
  #{ exec: "path-list" }
  #{ exec: "pci-ids" }
  #{ exec: "pgpass" }
  { exec: "pidstat" }
  { exec: "pidstat-s" }
  { exec: "ping" }
  { exec: "ping-s" }
  #{ exec: "pip-list" }
  #{ exec: "pip-show" }
  #{ exec: "pkg-index-apk" }
  #{ exec: "pkg-index-deb" }
  #{ exec: "plist" }
  #{ exec: "postconf" }
  #{ exec: "proc" }
  #{ exec: "ps" } # already in nu
  #{ exec: "resolve-conf" }
  #{ exec: "route" }
  #{ exec: "rpm-qi" }
  { exec: "rsync" }
  { exec: "rsync-s" }
  #{ exec: "semver" }
  { exec: "sfdisk" }
  #{ exec: "shadow" }
  #{ exec: "srt" }
  { exec: "ss" } 
  #{ exec: "ssh-conf" }
  #{ exec: "sshd-conf" }
  { exec: "stat" }
  { exec: "stat-s" }
  { exec: "swapon" }
  { exec: "sysctl" }
  #{ exec: "syslog" }
  #{ exec: "syslog-s" }
  #{ exec: "syslog-bsd" }
  #{ exec: "syslog-bsd-s" }
  { exec: "systemctl" }
  { exec: "systemctl-lj" }
  { exec: "systemctl-ls" }
  { exec: "systemctl-luf" }
  #{ exec: "systeminfo" }
  #{ exec: "time" }
  { exec: "timedatectl" }
  #{ exec: "timestamp" }
  #{ exec: "toml" }
  #{ exec: "top" }
  #{ exec: "top-s" }
  #{ exec: "tracepath" }
  #{ exec: "traceroute" }
  #{ exec: "tune2fs" }
  #{ exec: "udevadm" }
  #{ exec: "ufw" }
  #{ exec: "ufw-appinfo" }
  #{ exec: "uname" } # already in nu
  #{ exec: "update-alt-gs" }
  #{ exec: "update-alt-q" }
  #{ exec: "upower" }
  #{ exec: "uptime" } # already in nu as `sys host | get uptime`
  #{ exec: "url" } # already in nu
  #{ exec: "ver" }
  #{ exec: "veracrypt" }
  #{ exec: "vmstat" }
  #{ exec: "vmstat-s" }
  #{ exec: "w" }
  #{ exec: "wc" } # already in nu as `str stats`
  #{ exec: "wg-show" }
  #{ exec: "who" }
  #{ exec: "x509-cert" }
  #{ exec: "x509-csr" }
  #{ exec: "xml" }
  #{ exec: "xrandr" }
  #{ exec: "yaml" }
  #{ exec: "zipinfo" }
  { exec: "zpool-iostat" } # only on hypervisor
  { exec: "zpool-status" } # only on hypervisor
] | each {|c|
  let eargs = if ('eargs' in $c) { $c.eargs } else { "" } # necessary?
  let jcargs = if ('jcargs' in $c) { $c.jcargs } else { $"--($c.exec)" }
  let fname = if ('fname' in $c) { $c.fname } else { $c.exec }
  maybe-gen-function-jc $c.exec $file $eargs $jcargs $fname
}
```

</details>

<details>
  <summary>config.nu</summary>

``` nushell
# [...]

source aliases.nu

# [...]
```

</details>

I currently run it manually from time to time like this:
``` nushell
nu ~/.local/config/nushell/scripts/bin/gen-aliases.nu
```
It could also be done via the config, e.g. on each login and then make nu exec itself again.
